src/Program2.cs

```diff
@@ -0,0 +1,526 @@
+
+//        Console.Write("stu");
+//        yield return 44; // state 5, promise of a value is fulfilled, WaitForNext completes with true
+//                         // fourth call to TryGetNext returns true, moves to state -2
+//                         // fifth call to TryGetNext returns false
+//                         // call to WaitForNext returns false
+//    }
+
+//    Task Slow() => throw null;
+//}
+
+class Program
+{
+    static async Task Main()
+    {
+        IAsyncEnumerable<int> test = AsyncIterator();
+        IAsyncEnumerator<int> enumerable = test.GetAsyncEnumerator();
+
+        //foreach await (var value in enumerable)
+        //{
+        //    Console.WriteLine(value)
+        //}
+        while (await enumerable.WaitForNextAsync())
+        {
+            while (true)
+            {
+                int value = enumerable.TryGetNext(out bool success);
+                if (!success) goto outer_loop_continue;
+
+                Console.WriteLine(value);
+            }
+            outer_loop_continue:;
+        }
+    }
+
+    static IAsyncEnumerable<int> AsyncIterator()
+    {
+        var stateMachine = new Unprounouncable();
+        stateMachine.State = StateMachineStates.NotStartedStateMachine;
+        stateMachine.Builder = AsyncTaskMethodBuilder.Create();
+        stateMachine._valueOrEndPromise = new ManualResetValueTaskSourceLogic<bool>(stateMachine);
+        stateMachine._promiseIsActive = true;
+
+        // note: we don't start the machine.
+        return stateMachine;
+    }
+
+    // TODO what happens to exceptions? (should they bubble into value promise, or anywhere else?)
+    private sealed class Unprounouncable :
+        IAsyncStateMachine,
+        IAsyncEnumerable<int>,
+        IAsyncEnumerator<int>,
+        IValueTaskSource<bool>, // used as the backing store behind the ValueTask<bool> returned from each MoveNextAsync
+        IStrongBox<ManualResetValueTaskSourceLogic<bool>> // exposes its ValueTaskSource logic implementation
+    {
+        // If you need to check for states other than NotStarted or Finished, then you must first ensure the promise is not set (ie. _promiseIsActive is false). Otherwise, the machine may be running on another thread.
+        public int State;
+
+        // awaiter and builder are used for 'await' in the method body
+        public TaskAwaiter Awaiter; // thin wrapper around a Task (that we'll be waiting on)
+        public AsyncTaskMethodBuilder Builder; // used for getting a callback to MoveNext when async code completes
+
+        private int _current;
+
+        #region promise
+        // All of the logic for managing the IValueTaskSource implementation
+        public ManualResetValueTaskSourceLogic<bool> _valueOrEndPromise; // promise for a value or end (true means value found, false means finished state)
+        public bool _promiseIsActive = false; // this is spiritually equivalent to saying the promise is set versus null
```

Comment:
@alekseyts You were wondering whether we could dispense with this bool value, and maybe encode it in the state.
I'd like to flesh out why I don't think that is possible:
```C#
label:
Write("A");
yield return 42; // state 1 (before the value was returned by TryGetNext) and state 2 (afterwards)
```
If there is any code that jumps to the label, it necessarily must reach state 1. When reaching a suspension point, it is that suspension point that determines the state (`this.State = <some value>;`). The previous state currently has no say in that.
If we wanted to have the previous state play a role, I think we'd have to save it. You could argue that's already what this bool does (it save the essential information from previous state).

That said, I do wonder if such a bool could be encoded into `ManualResetValueTaskSourceLogic` (maybe converting it's bool field to be an enum type instead? Tagging @stephentoub

Reply:
It seems like you need the bool to be able to differentiate whether TryGetNext is being called after a call to WaitForNextAsync (in which case it should use the value yielded in that operation) or after a call to TryGetNext (in which case it should MoveNext).  Given that, tracking the state separately from the MRVTSL makes sense to me, as it's not logically part of the WaitForNextAsync ValueTask.